src/domain/mongoose_domain_loader.erl

```diff
@@ -0,0 +1,82 @@
+-module(mongoose_domain_loader).
+-export([load_data_from_base/2]).
+-export([check_for_updates/2]).
+
+-include("mongoose_logger.hrl").
+
+load_data_from_base(FromNum, PageSize) ->
```

Comment:
Yeah, I was thinking about that. There are some issues:
- Routing to unknown domain would become slower.
- Cache invalidation? Maybe nope. rpc could fail, updates could come from a node, which is net-splited from the main cluster.
- The cost of not applying some change is extremely high. Basically, the whole tenant would be not available.
- After the full restart we would have different performance characteristics, because the ETS table is cold.
- By the way, 10k requests at once to a single cold domain - totally possible, we would have to have logic to track loading. Which is also code, just not on the DB side ;) 
- We can still do lazy init for domains, if we need to do init for domains (like starting some gen_servers).
- Time of execution of get_host_type becomes random, which could break some time-sensitive code.

That is something similar to MUC rooms.
As long as there is less than 1M domains, it makes sense to store the whole set on MIM side.
Once we want 100M domains - we would have to do lazy loading (which is not the case for us).


My main considerations:
- to have DB to be the source of truth.
- to never miss disable/enable/remove/insert of a domain.
- To not use erlang distribution for anything critical. 
- To not use Mnesia.
- Once node starts, it is ready to work as fast as possible.
- If DB is not available on any of the nodes - we would miss the updates, but once it's back - MIM recovers easily.

Reply:
Thanks for sharing the rationale behind your solution, this was the main goal of my comment :-)

I think that relying on periodic updates is risky as we are deliberately making the data out of sync for some time, even on the very node that performs the change (albeit for a shorter period). When routing the message to a different node, we might get different domain mapping for many stanzas which might cause a flood of errors (for some modules) until the eventual consistency is reached. I think that preventing such errors is non-trivial.

Periodic events also make bugs more random and difficult to reproduce. In all systems with multiple periodic events there was always a tricky bug caused by an unlikely ordering of such events. With more complexity there is always more opportunities for bugs to creep in.

Reply:
>     * We can still do lazy init for domains, if we need to do init for domains (like starting some gen_servers).
I don't think any `gen_server` needs to be started per domain. The only values initialized per domain would be very simple like the mapping from a MUC subdomain to the host type. It can be either lazily stored (cached) in an ETS or just replaced with a simple bit of code that calculates it dynamically like I did for Motorola. It had no  impact on load test results.

Reply:
Each time we change a domain, we notify all nodes, that there is something new in the database.
So, it speeds up the update from the default 0-30 seconds, till the domain is updated.

But there is no requirement to block the insert_domain request, until the operation has been applied across all nodes, so I haven't added it.

But if you want to block until it's applied across all nodes, just call `service_domain_db:sync()`, it will block until everything is consumed from the DB. We use this function in our tests.